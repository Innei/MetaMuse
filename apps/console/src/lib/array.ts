export function arrayMoveMutable(
  array: unknown[],
  fromIndex: number,
  toIndex: number,
) {
  const startIndex = fromIndex < 0 ? array.length + fromIndex : fromIndex

  if (startIndex >= 0 && startIndex < array.length) {
    const endIndex = toIndex < 0 ? array.length + toIndex : toIndex

    const [item] = array.splice(fromIndex, 1)
    array.splice(endIndex, 0, item)
  }
}

export function arrayMoveImmutable<ValueType>(
  array: readonly ValueType[],
  fromIndex: number,
  toIndex: number,
): ValueType[] {
  const newArray = [...array]
  arrayMoveMutable(newArray, fromIndex, toIndex)
  return newArray
}
