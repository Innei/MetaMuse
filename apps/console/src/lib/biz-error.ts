import type { ErrorCodeEnum } from '@constants'
import type { AxiosError } from 'axios'

export class BizError extends Error {
  constructor(
    public readonly code: ErrorCodeEnum,
    public readonly message: string,

    public readonly status: number,
    public readonly raw: AxiosError,
  ) {
    super()
  }
}
