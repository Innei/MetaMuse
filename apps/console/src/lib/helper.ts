import clsx from 'clsx'
import { twMerge } from 'tailwind-merge'

export const clsxm = (...args: any[]) => {
  return twMerge(clsx(args))
}

export const sleep = (ms: number) => {
  return new Promise((resolve) => setTimeout(resolve, ms))
}
