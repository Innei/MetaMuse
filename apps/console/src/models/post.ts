import { Category, Post, PostTag } from '@model'

export type PostModel = Post & {
  category: Category
  tags: PostTag[]
}
