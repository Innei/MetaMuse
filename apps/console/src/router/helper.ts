import { useCallback, useMemo } from 'react'
import type { RouteMeta } from './interface'

import { useI18n } from '~/i18n/hooks'

export const useExtractTitle = (maybeTitle: RouteMeta['title'] | undefined) => {
  const t = useI18n()
  return useMemo(
    () => (typeof maybeTitle === 'function' ? maybeTitle(t) : maybeTitle),
    [maybeTitle, t],
  )
}

export const useExtractTitleFunction = () => {
  const t = useI18n()
  return useCallback(
    (maybeTitle: RouteMeta['title'] | undefined) =>
      typeof maybeTitle === 'function' ? maybeTitle(t) : maybeTitle,
    [t],
  )
}
