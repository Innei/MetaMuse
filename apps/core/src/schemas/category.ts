import { createZodDto } from 'nestjs-zod/dto'
import * as z from 'nestjs-zod/z'

import { CompletePost, RelatedPostModel } from './index'

export const CategoryModel = z.object({
  id: z.string(),
  name: z.string(),
  slug: z.string(),
  created: z.date(),
})

export class CategoryDto extends createZodDto(CategoryModel) {}

export interface CompleteCategory extends z.infer<typeof CategoryModel> {
  Post: CompletePost[]
}

/**
 * RelatedCategoryModel contains all relations on your model in addition to the scalars
 *
 * NOTE: Lazy required in case of potential circular dependencies within schema
 */
export const RelatedCategoryModel: z.ZodSchema<CompleteCategory> = z.lazy(() =>
  CategoryModel.extend({
    Post: RelatedPostModel.array(),
  }),
)
