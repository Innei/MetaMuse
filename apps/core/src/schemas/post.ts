import { createZodDto } from 'nestjs-zod/dto'
import * as z from 'nestjs-zod/z'

import { CompleteCategory, RelatedCategoryModel } from './index'

export const PostModel = z.object({
  id: z.string(),
  slug: z.string(),
  text: z.string(),
  title: z.string(),
  created: z.date(),
  modified: z.date().nullish(),
  categoryId: z.string(),
})

export class PostDto extends createZodDto(PostModel) {}

export interface CompletePost extends z.infer<typeof PostModel> {
  category: CompleteCategory
}

/**
 * RelatedPostModel contains all relations on your model in addition to the scalars
 *
 * NOTE: Lazy required in case of potential circular dependencies within schema
 */
export const RelatedPostModel: z.ZodSchema<CompletePost> = z.lazy(() =>
  PostModel.extend({
    category: RelatedCategoryModel,
  }),
)
