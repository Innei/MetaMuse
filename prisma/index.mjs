import './global'

export * from './client'
export * from './zod/modelSchema'
