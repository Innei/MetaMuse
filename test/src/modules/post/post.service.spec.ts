import { PostService } from '@core/modules/post/post.service'
import { Prisma } from '@prisma/client'
import { createServiceUnitTestApp } from '@test/helper/create-service-unit'
import { prisma } from '@test/lib/prisma'
import { generateMockCategory } from '@test/mock/data/category.data'
import { generateMockPost, mockPostInputData } from '@test/mock/data/post.data'
import { mockedEventManagerService } from '@test/mock/helper/helper.event'

describe('/modules/post/post.service', () => {
  const proxy = createServiceUnitTestApp(PostService)

  const createMockCategory = async () => {
    return await prisma.category.create({
      data: {
        ...generateMockCategory(),
      },
    })
  }

  it('should create post successful', async () => {
    const category = await createMockCategory()
    const { id } = category

    const result = await proxy.service.create({
      ...mockPostInputData,
      categoryId: id,
    })

    expect(result).toMatchObject({
      ...mockPostInputData,
      categoryId: id,
      category,
    })
    expect(mockedEventManagerService.event).toBeCalledWith(
      'POST_CREATE',
      result,
    )
  })

  it('should throw when post exist', async () => {
    const category = await createMockCategory()
    await prisma.post.create({
      data: {
        ...mockPostInputData,
        categoryId: category.id,
      },
    })
    const { id } = category
    await expect(
      proxy.service.create({
        ...mockPostInputData,
        categoryId: id,
      }),
    ).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should throw when category not found', async () => {
    await expect(
      proxy.service.create({
        ...mockPostInputData,
        categoryId: 'not-found',
      }),
    ).rejects.toThrowErrorMatchingSnapshot()
  })

  it('should paginate posts successful', async () => {
    const cate = await createMockCategory()
    const mockedDataList = [] as Prisma.PostCreateManyInput[]
    for (let i = 0; i < 20; i++) {
      mockedDataList.push({
        ...generateMockPost(),
        categoryId: cate.id,
      })
    }

    await prisma.post.createMany({
      data: mockedDataList,
    })

    const pagination = await proxy.service.paginatePosts({
      page: 1,
      size: 5,
    })

    expect(pagination.pagination).toMatchSnapshot()
    expect(pagination.data[0]!.category).toMatchObject(cate)
  })

  it('should get post by id successful', async () => {
    const cate = await createMockCategory()
    const post = await prisma.post.create({
      data: {
        ...generateMockPost(),
        categoryId: cate.id,
      },
    })

    const result = await proxy.service.getPostById(post.id)

    expect(result).toMatchObject(post)
  })

  it('should get post throw when 404', async () => {
    expect(
      proxy.service.getPostById('not-found'),
    ).rejects.toThrowErrorMatchingSnapshot()
  })
})
