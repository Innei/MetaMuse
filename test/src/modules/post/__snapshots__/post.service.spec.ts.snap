// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/modules/post/post.service > should get post throw when 404 1`] = `"post not found"`;

exports[`/modules/post/post.service > should throw when category not found 1`] = `"category not found"`;

exports[`/modules/post/post.service > should throw when post exist 1`] = `"post already exist"`;
